p2panda-net-next/src/actors/network.rs

```diff
@@ -0,0 +1,245 @@
+            Some("endpoint".to_string()),
+            Endpoint {},
+            (),
+            myself.clone().into(),
+        )
+        .await?;
+
+        // Spawn the address book actor.
+        let (address_book_actor, _) = Actor::spawn_linked(
+            Some("address book".to_string()),
+            AddressBook {},
+            (),
+            myself.clone().into(),
+        )
+        .await?;
+
+        // Spawn the discovery actor.
+        let (discovery_actor, _) = Actor::spawn_linked(
+            Some("discovery".to_string()),
+            Discovery {},
+            (),
+            myself.clone().into(),
+        )
+        .await?;
+
+        let state = NetworkState {
+            events_actor,
+            events_actor_failures: 0,
+            endpoint_actor,
+            endpoint_actor_failures: 0,
+            address_book_actor,
+            address_book_actor_failures: 0,
+            discovery_actor,
+            discovery_actor_failures: 0,
+        };
+
+        Ok(state)
+    }
+
+    async fn post_start(
+        &self,
+        _myself: ActorRef<Self::Msg>,
+        _state: &mut Self::State,
+    ) -> Result<(), ActorProcessingErr> {
+        Ok(())
+    }
+
+    async fn post_stop(
+        &self,
+        _myself: ActorRef<Self::Msg>,
+        _state: &mut Self::State,
+    ) -> Result<(), ActorProcessingErr> {
+        Ok(())
+    }
+
+    async fn handle(
+        &self,
+        _myself: ActorRef<Self::Msg>,
+        _message: Self::Msg,
+        _state: &mut Self::State,
+    ) -> Result<(), ActorProcessingErr> {
+        Ok(())
+    }
+
+    async fn handle_supervisor_evt(
+        &self,
+        myself: ActorRef<Self::Msg>,
+        message: SupervisionEvent,
+        state: &mut Self::State,
+    ) -> Result<(), ActorProcessingErr> {
+        match message {
+            SupervisionEvent::ActorStarted(actor) => {
+                if let Some(name) = actor.get_name() {
+                    debug!("network actor: received ready from {} actor", name);
+                }
+            }
+            SupervisionEvent::ActorFailed(actor, panic_msg) => {
+                match actor.get_name().as_deref() {
+                    Some("events") => {
+                        warn!("network actor: events actor failed: {}", panic_msg);
+
+                        // Respawn the events actor.
+                        let (events_actor, _) = Actor::spawn_linked(
+                            Some("events".to_string()),
+                            Events {},
+                            (),
+                            myself.clone().into(),
+                        )
+                        .await?;
+
+                        state.events_actor_failures += 1;
+                        state.events_actor = events_actor;
+                    }
+                    Some("endpoint") => {
+                        warn!("network actor: endpoint actor failed: {}", panic_msg);
```

Comment:
What causes an endpoint actor to fail?

Reply:
Good question. I don't foresee any failure cases in the endpoint actor. There are some errors which could emerge when initialising the actor (during the `pre_start()` method) but those would not invoke the supervision strategy. So for now it's fair to say that this failure check is "just in case" and could help with debugging unexpected behaviour.